File: nezal-ar/app/models/games/base_defender/weapon.rb
module BD 

  class Weapon

    @name = "weapons"
    @building  = BD::Wedge.name

    def self.buy(user_game_profile, wedge_coords, weapon)
      game_data = Game::current.data
      location_hash = BaseDefender.convert_location(wedge_coords)
      
      validation = validate(user_game_profile, wedge_coords)
      return validation if validation['valid'] == false

      # Get selected weapon data from metadata.
      weapon_data = game_data[@name][weapon]
      # Add the weapon to the selected building.
      user_game_profile.data[@building][location_hash][@name] = weapon
      # Remove the cost from of the weapon from the user coins/resources 
      user_game_profile.user.coins -= weapon_data['cost']['coins'] if weapon_data['cost']['coins']
      ['rock', 'lumber'].each do | key | 
        if weapon_data['cost'][key]
          user_game_profile.data[key] -= weapon_data['cost'][key]
        end
      end
      return validation
    end

    def self.sell(user_game_profile, wedge_coords, weapon)
    end

    def self.validate(user_profile, coords, weapon)
      game_data = Game::current.data
      # Make sure there is a building at the described location.
      if (!user_profile.data[@building][coords]) || 
          (user_profile.data[@building][coords]['level'] < 1)
        return {'valid' => false,
                'error' => "There is no wedge at the selected location, OR wedge still under construction." }
      end
      # Make sure there is enough coins/resources to cover the cost of action.
      validation = {'valid' => true, 'error' => "Not enough resources, you need more " }
      neededCoins = game_data[@name][weapon]['cost']['coins']
      if( neededCoins  && (neededCoins - user_profile.user.coins) > 0)
        validation['valid'] = false
        validation['error'] += (neededCoins - user_profile.user.coins) + " coins .. "
      end
      neededRock = game_data[@name][weapon]['cost']['rock']
      if( neededRock && (neededRock - user_profile.rock) > 0)
        validation['valid'] = false
        validation['error'] += (neededRock - user_profile.rock) + " rocks .. "
      end      
      neededLumber = game_data[@name][weapon]['cost']['lumber']
      if( neededLumber &&  (neededLumber - user_profile.lumber) > 0)
        validation['valid'] = false
        validation['error'] += (neededLumber - user_profile.lumber) + " lumber .. "
      end
      return validation
    end

    attr_accessor :owner, :coords, :angle, :attacker, :rocks, :rock_num, :specs
    # Create a new weapon instance for the wedges to detect the attack
    # Should take the wedge map building as owner to simulate the attack
    def initialize(owner, creeps)
      @creeps = creeps 
      @owner = owner
      @attacker = nil
      @step = 0
      @tick = 0
      @angle = 0
      @rocks = []
      @weapon = Game::current.buildings['wedge']['levels'][@owner.owner['level'].to_s]['weapon']
      @specs = Game::current.buildings['wedge']['levels'][@owner.owner['level'].to_s]['specs']
      @coords = @owner.owner['coords']
      @animated = false
      @move_steps = 2
    end

    def tick      
      if(@owner.owner['state'] == BD::Building.states['NORMAL'])
        check_attack()
        if @attacker
          @angle = Map.get_general_direction(@coords['x'], @coords['y'], @attacker.coords['x'], @attacker.coords['y'])
        end 
#        puts "Weapon Tick :: #{self.__id__} "
      end 
    end

    def display_tick
      if(@owner.owner['state'] == BD::Building.states['NORMAL'])
        if @attacker == nil
          @step = 0
          @animated = false
        elsif @attacker && @animated == false
          @animated = true
        end
        if @attacker && @animated
          @step += 1
          if @step == @move_steps
            rock = BD::Rock.new(self, @attacker)
            @rocks << rock
          end
          if @step == 9
            @step = 0
            @animated = false
            fire
          end
        end
#        puts "Weapon Render :: #{self.__id__} :: #{@step} :: #{@animated}"
      end
      @rocks.each do |rock| 
                    if(!rock.done) 
                      rock.tick 
                    end
                  end
    end

    def check_attack()
      if(@owner.hp <= 1)
        @attacker = nil 
        return
      end
      if (@attacker && @attacker.hp > 0)
        return
      end
      attack = nil
      minHp = 50000
      minDistance = @specs['range']
      @creeps.each do |creep|
        if creep.hp > 0 
          dist = Util.distance(@coords['x'], @coords['y'], creep.coords['x'], creep.coords['y']);
          if(dist < minDistance)
            if(creep.hp < minHp)
              minHp = creep.hp;
              attack = creep;
            end
          end
        end
      end
      @attacker = attack;
    end

    def fire 
      @attacker = nil
    end

  end

end

File: nezal-ar/public/base-defender/js/game/weapon.js
var Weapon = Class.create({

  angle : 0,
  targetAngle : 5,
  imgHeight : 93,
  imgWidth : 64,
  
  name : null,
  game : null,
  specs : null,
  attacked : false,

  initialize : function(name, owner) {
    this.name = name;
    this.coords = owner.coords;
    this.game = owner.game;
    this.owner = owner;
    var newAngle = Math.round(Math.random() * 10 );
    this.angle = newAngle % 8;
    this.specs = this.game.data.weapons[this.name].specs;
		this.game.scene.push(this);
    var self = this;
  	this.game.scene.pushPeriodicalRenderLoop(1,1,2,function() {self.randomDirectionChange()});
  	this.game.scene.pushPeriodicalRenderLoop(2,4,1,function() {self.changeAngle()});
    this.rock = new Rock(this);
  },

  tick : function() {
    if(this.owner.state == this.owner.states.NORMAL)
    {
      this.attacker = this.checkAttack();
      if ( this.attacker ) {
        this.attacked = true;
        this.angle = Map.getGeneralDirectoin(this.coords.x, this.coords.y, this.attacker.coords.x, this.attacker.coords.y);
      }else {
        this.attacked = false;
      }
    }
  },

  checkAttack : function() {
    if(this.owner.hp <=1 || this.attacker)
      return;
    var attack = null;
    var minHp = 50000;
    var minDistance = 150;
    for( var i = 0 ; i < this.game.creepFactory.registery.length; i++ )
    {
      var creep = this.game.creepFactory.registery[i];
      var dist = Util.distance(this.coords.x, this.coords.y, creep.coords.x, creep.coords.y);
      if(dist < minDistance)
      {
        if(creep.hp < minHp)
        {
          minHp = creep.hp;
          attack = creep;
        }
      }
    }
    return attack;
  },

  fire : function() {
    if( this.owner.hp <=1 )
    {
      this.attacked = false;
      this.attacker = null;
      return;
    }      
    if(this.attacker)
      this.attacker.hp -= this.specs.power;
    this.attacked = false;
    this.attacker = null;
  },
  
  randomDirectionChange : function() {
    if( this.owner.state == this.owner.states.NORMAL && this.attacked == false ) {
      var newAngle = Math.round(Math.random() * 10 );
      this.targetAngle = newAngle % 8;
    }
  },

  changeAngle : function() {
    if( this.owner.state == this.owner.states.NORMAL && this.attacked == false ) {
		  if(this.angle > this.targetAngle){
			  if((this.targetAngle+8) - this.angle < this.angle - this.targetAngle){
				  this.angle = (this.angle+1) % 8 
			  }else{
				  this.angle = (this.angle+7) % 8
			  }
		  }else if (this.angle < this.targetAngle){
			  if(this.targetAngle - this.angle < (this.angle+8) - this.targetAngle){
				  this.angle = (this.angle+1) % 8 
			  }else{
				  this.angle = (this.angle+7) % 8
			  }
		  }			
    }
  }


});
